src/v/storage/readers_cache.cc

```diff
@@ -278,19 +279,27 @@ readers_cache::dispose_entries(intrusive_list<entry, &entry::_hook> entries) {
 
 void readers_cache::dispose_in_background(
   intrusive_list<entry, &entry::_hook> entries) {
-    try {
-        (void)ss::with_gate(
+        ssx::spawn_with_gate(
           _gate, [this, entries = std::move(entries)]() mutable {
               return dispose_entries(std::move(entries));
           });
-    } catch (const ss::gate_closed_exception& e) {
-        vlog(stlog.debug, "gate closed while disposing reader");
-    }
 }
 
 void readers_cache::dispose_in_background(entry* e) {
-    try {
-        (void)ss::with_gate(_gate, [this, e] {
+    if (_gate.is_closed()) {
+        /**
+         * since gate is closed and we failed to call finally on the reader we
+         * add it back to _readers list, it will be gracefully closed in
+         * `readers_cache::close`.
+         * NOTE:
+         * _readers intrusive list is supposed to keep resusable readers but
+         * when gate closed exception is thrwon we are certain that no more
+         * oprations will be executed on the cache so we can reuse the _readers
+         * list to gracefully shutdown readers.
+         */
+        _readers.push_back(*e);
+    } else {
+        ssx::spawn_with_gate(_gate, [this, e] {
```

Comment:
I think this transformation brings up a good semantic that might be worth mentioning along side `spawn_with_gate`'s implementation which is that it should preserve the property that the continuation is invoked atomically with checking the that the gate is closed. ie, no scheduling points at the beginning of spawn_with_gate.